sourcegraph/zoekt
# sourcegraph: multi-tenant zoekt

This updates webserver and sourcegraph-indexserver to support multi-tenancy.

The change is behind an ENV feature-flag. Apart from changes to the grpc message format for `IndexOptions` and `ListResponse` (-> requires update in Sourcegraph), this change is a noop if multi-tenancy is not enabled. Especially, other users of Zoekt should not be affected.

Key changes:
- sourcegraph-indexserver
  - The list call now returns a map "tenant id -> list of repo ids". Before, we returned just a list of repo ids.
  - Iterate over map to get index options per tenant
  - Set index dir per tenant
  - Note: Changes are limited to sourcegraph-indexserver. "gitindex" and "builder" are not affected.
- webserver
  - watcher: watch data/index/tenants for new tenant dirs 
  - check if path of index matches tenant

Notes:
- This design will not scale well beyond a hundred tenants, mostly because we iterate over all tenants to get `IndexOptions`. For better performance we might have to switch the communication and let Sourcegraph push index jobs to Zoekt.
- The debug pages need more thought. I had disabled them initially but in this PR they are unchanged. Things like the reindex button on the indexserver debug page will not work, because we don't have the tenant information.

Future:
- We might consider skipping entire tenant directories instead of checking shard by shard
- Enable shard merging per tenant
- Make "force reindex" work if the requests comes via the socket connection

Test plan:
- new units tests
- manual testing: 
  - I ran this PR together with a multi-tenant instance of Sourcegraph and confirmed that I can run indexed search per tenant as expected.
  - I confirmed backward compatibility by running this PR against a Sourcegraph instance without multi-tenancy.
